Configuration API

Owen1212055 opened the pull request:
Please provide feedback on naming, and general event consensis. 

But my goal is to have the following:
- Full cookie getting/setting support in the login/config/play stage. This means you'll be able to disconnect players during LOGIN depending on sent cookies. (Bukkit only allows play stage, and PARTLY config stage) (`CookieConnection`)
- Allowing to boot players into the config stage/ out of it
- Allow to add custom configuration tasks, so for example being able to load player data. (`AsyncPlayerConnectionConfigurateEvent`)
- Allowing to load resource packs during config stage better
- Transfer the player in config/player
- send custom report details (shown in client debug log)

Owen1212055 committed:
WIP Configuration API

Fix disconnect continuing to tick ServerLoginPacketListenerImpl

Fixes https://github.com/PaperMC/Paper/issues/8676

This implements a solution that preemptively exits the tick loop if we have already marked the connection as disconnecting. This avoids changing the result of Connection#isConnected in order to avoid other possibly unintentional changes. Fundamentally it should be investigated if closing the connection async is really still needed.

This also additionally removes the login disconnecting logic for server stopping, as this was also a possible issue in the config stage but also shouldn't be an issue as connections are closed on server stop very early.

Additionally, do not check for isConnecting() on VERIFYING, as that seemed to be an old diff originally trying to resolve this code, however isConnected is not updated at this point so it pretty much was useless.

lynxplay commented on paper-api/src/main/java/io/papermc/paper/connection/CookieConnection.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.concurrent.CompletableFuture;
+
+@NullMarked
+public interface CookieConnection extends PlayerConnection {
```
PlayerCookieConnection

lynxplay replied:
Experimental the living shit out of *all* of these types.

lynxplay commented on paper-api/src/main/java/io/papermc/paper/connection/PlayerConfigurationConnection.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @return The IP address
+     */
+    InetAddress getAddress();
```
This is kinda doubled between this and the login connection and then it isn't present on the game connection because player presumably has it.

While the profile may not be available, address information (Inet or Socket (hi emily)) feels very very basic?

Owen1212055 replied:
Abstracted out. Thanks for this

lynxplay commented on paper-api/src/main/java/io/papermc/paper/connection/PlayerConfigurationConnection.java:
```diff
@@ -0,0 +1,58 @@
+    InetAddress getAddress();
+
+    /**
+     * Gets the raw address of the player logging in
```
Dots at the end of jd bodies.
Goes for the rest of this PR too.

lynxplay commented on paper-api/src/main/java/io/papermc/paper/connection/PlayerConfigurationConnection.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * Completes the configuration for this player, which will cause this player to reenter the game.
+     * <p>
+     * Note, this should be only be called if you are reconfiguring the player.
```
This, similar to #enterConfiguration poses the interesting issue of "ah, you just requested what is basically a destruction of this connection state, what happens to *this* instance".

Calling #clearChat after this is illegal.
Getters *might* be fine. But we should look into a proper way to signal, "you destroyed this connection instance".

Owen1212055 replied:
I agree here. But, I am not sure as the current standard has been to silently fail in many places.
![image](https://github.com/user-attachments/assets/45d40776-2a02-47d9-8d7d-9cc1a8271a31)
So we could very much add some validation, but it all ends up nooping.

lynxplay commented on paper-api/src/main/java/io/papermc/paper/connection/PlayerConnection.java:
```diff
@@ -0,0 +1,24 @@
+    void disconnect(Component component);
+
+    /**
+     * Gets if this connection originated from a transferred connection.
```
Formatting, obligatory note that this is completely controlled by the client and to be taken with a grain of salt.

lynxplay commented on paper-api/src/main/java/io/papermc/paper/connection/PlayerLoginConnection.java:
```diff
@@ -0,0 +1,58 @@
+import java.net.InetSocketAddress;
+
+@NullMarked
+public interface PlayerLoginConnection extends CookieConnection  {
```
Extra useless space.

lynxplay commented on paper-api/src/main/java/io/papermc/paper/connection/PlayerLoginConnection.java:
```diff
@@ -0,0 +1,58 @@
+     * @return The IP address
+     */
+    @NotNull
+    InetAddress getAddress();
```
hi emily

Owen1212055 replied:
idk what this means? ;-;

lynxplay commented on paper-api/src/main/java/io/papermc/paper/event/connection/configuration/AsyncPlayerConnectionConfigureEvent.java:
```diff
@@ -0,0 +1,33 @@
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A task that allows you to configurate the player.
```
This is an event, not a task.

lynxplay commented on paper-api/src/main/java/io/papermc/paper/event/connection/configuration/AsyncPlayerConnectionConfigureEvent.java:
```diff
@@ -0,0 +1,33 @@
+ * <p>
+ * This occurs after configuration, but before the player has entered the world.
+ */
+public class AsyncPlayerConnectionConfigureEvent extends PlayerConfigurationConnectionEvent {
```
Naming here indeed hits the wall a bit.
A "AsyncPlayerConnectionConfigureEvent", representing when a player is near-finished with their configuration phase, feels rough, the parent name "PlayerConfigurationConectionEvent" makes this even worse, either it is Configuration or Configure, pick one xD
The fact they are swapped also hurts a bit.

lynxplay commented on paper-api/src/main/java/io/papermc/paper/event/connection/configuration/PlayerConnectionInitialConfigureEvent.java:
```diff
@@ -0,0 +1,27 @@
+import org.jetbrains.annotations.NotNull;
+
+public class PlayerConnectionInitialConfigureEvent extends PlayerConfigurationConnectionEvent {
+    private static final HandlerList handlers = new HandlerList();
```
Here and everywhere else, HANDLERS, or HANDLER_LIST.

lynxplay commented on paper-server/patches/sources/net/minecraft/network/DisconnectionDetails.java.patch:
```diff
@@ -0,0 +1,17 @@
+ import net.minecraft.network.chat.Component;
+ 
+-public record DisconnectionDetails(Component reason, Optional<Path> report, Optional<URI> bugReportLink) {
++// Paper start
```
identifying comments. What API is this for.

lynxplay commented on paper-server/patches/sources/net/minecraft/server/network/ServerCommonPacketListenerImpl.java.patch:
```diff
@@ -1,11 +1,11 @@
  
 -public abstract class ServerCommonPacketListenerImpl implements ServerCommonPacketListener {
-+public abstract class ServerCommonPacketListenerImpl implements ServerCommonPacketListener, org.bukkit.craftbukkit.entity.CraftPlayer.TransferCookieConnection { // CraftBukkit
++public abstract class ServerCommonPacketListenerImpl implements ServerCommonPacketListener{
```
Useless diff.

Machine-Maker commented on paper-api/src/main/java/io/papermc/paper/event/connection/configuration/PlayerConfigurationConnectionEvent.java:
```diff
@@ -0,0 +1,24 @@
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Internal
+abstract class PlayerConfigurationConnectionEvent extends Event {
```
We don't need this abstract event. Just pull the one field up to the parent events.

Machine-Maker commented on paper-server/patches/sources/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java.patch:
```diff
@@ -46,19 +70,38 @@
++            if (this.connection.savedPlayerForLoginEventLegacy != null) {
++                playerForLogin = this.connection.savedPlayerForLoginEventLegacy;
++                // WE have to do this because the player isnt updated properly
++                playerForLogin.updateOptions(this.clientInformation);
```
calling updateOptions will fire the optionschange event. Which makes that event now fire sometimes and not other times depending on if anyone is listening to PlayerLoginEvent

use updateOptionsNoEvents I think.

Machine-Maker commented on paper-server/patches/sources/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java.patch:
```diff
@@ -70,3 +113,44 @@
+     }
+ 
+-    private void finishCurrentTask(ConfigurationTask.Type taskType) {
++    public void finishCurrentTask(ConfigurationTask.Type taskType) { // Paper
```
at

Machine-Maker commented on paper-server/patches/sources/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java.patch:
```diff
@@ -1,18 +1,29 @@
+     private static final Logger LOGGER = LogUtils.getLogger();
+     private static final Component DISCONNECT_REASON_INVALID_DATA = Component.translatable("multiplayer.disconnect.invalid_player_data");
+-    private final GameProfile gameProfile;
++    public final GameProfile gameProfile; // Paper
```
at

Machine-Maker commented on paper-server/patches/sources/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java.patch:
```diff
@@ -1,18 +1,29 @@
+-    private ConfigurationTask currentTask;
+-    private ClientInformation clientInformation;
++    public ConfigurationTask currentTask; // Paper
++    public ClientInformation clientInformation; // Paper
```
at

Machine-Maker commented on paper-api/src/main/java/io/papermc/paper/event/player/PlayerServerFullCheckEvent.java:
```diff
@@ -0,0 +1,100 @@
+ * trying to join the server while it is in this state.
+ */
+@NullMarked
+public class PlayerServerFullCheckEvent extends Event {
```
Wouldn't this event be more useful if functioned as a possible "override" for the entire fullnest check instead of only firing when the server was full? We can use the event tristate for it, and have a method that returns whether vanlla logic would allow or deny.

Machine-Maker commented on paper-api/src/main/java/io/papermc/paper/connection/PlayerConnection.java:
```diff
@@ -0,0 +1,20 @@
+
+import net.kyori.adventure.text.Component;
+
+public interface PlayerConnection {
```
It would be really good if the hierarchy of types here was adjusted so this class could be sealed and permit the two API interfaces. Would make switching on it nicer, especially in events where you only have the PlayerConnection, and to know what type it is, you do a switch pattern match. It's not a super simple change, because of the impl CommonCookeConnection abstract class, but things could def be moved around.

Owen1212055 replied:
After playing around with this a bit, I am not sure about it really. This does require alot of moving around and possible duplication that I am not fully sure is worth it.

Additionally, its a bit of an odd situation, as I am not sure in what cases you'd be casting like this in the first place.

Machine-Maker commented on paper-api/src/main/java/io/papermc/paper/connection/PlayerCommonConnection.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Represents a connection that has properties shared between the GAME and CONFIG stage.
+ */
+public interface PlayerCommonConnection extends CookieConnection {
```
Same situation here regarding sealed interfaces being nicer for switch pattern matching (doesn't require default case)

Owen1212055 replied:
see above

Machine-Maker commented on paper-api/src/main/java/io/papermc/paper/event/connection/common/PlayerFinishConnectionPhaseEvent.java:
```diff
@@ -0,0 +1,80 @@
+ * <p>
+ * Currently fires during the login and configuration protocol.
+ */
+public class PlayerFinishConnectionPhaseEvent extends Event {
```
Isn't this event a little redundant? It seems to just fire right after the PlayerConnectionValidateLoginEvent. Any way these can be condensed into one? If not, the javadoc comments probably need to talk about the difference between the two.

Owen1212055 replied:
Yep, I think you are correct.

Owen1212055 replied:
See https://discord.com/channels/289587909051416579/925530366192779286/1385780483815637042

Machine-Maker commented on paper-api/src/main/java/io/papermc/paper/event/connection/configuration/PlayerConnectionReconfigurateEvent.java:
```diff
@@ -0,0 +1,28 @@
+ * Indicates that this player is being reconfigured, meaning that this connection will be held in the configuration
+ * stage unless kicked out through {@link PlayerConfigurationConnection#completeConfiguration()}
+ */
+public class PlayerConnectionReconfigurateEvent extends PlayerConfigurationConnectionEvent {
```
I feel like this event, and PlayerConnectionInitialConfigureEvent can be compiled with just an enum "cause" or something. They are just both fired right before switching to configuration phase, from either login or from game. Just a PlayerEnterConfigurationPhaseEvent or something with a cause?

Owen1212055 replied:
The reason why I made this separate is because both are treated very differently. 
Basically I felt like differenciating that this player is going to be held forever in config phase vs they are going through the config phase as normal was an important enough disctinction.

Machine-Maker replied:
But in both cases, doesn't the player just automatically go through the tasks in the configuration phase?

In plain vanilla, if you call that function to kick the player back to config phase, won't they just come right back in as if they had just logged in? Or do you have to somehow trigger something to *let* them back in?

Owen1212055 replied:
No, in reconfiguration the tasks are not run. This causes the player to just be stuck in the stage until released.
startConfiguration is only called when login packet is received, so there are no "tasks" that are being executed.

You have to "somehow" trigger to let them back in.

Owen1212055 replied:
See https://discord.com/channels/289587909051416579/925530366192779286/1385790639345631333

kennytv commented on paper-api/src/main/java/io/papermc/paper/connection/PlayerConfigurationConnection.java:
```diff
@@ -0,0 +1,50 @@
+     * <p>
+     * Note, this should be only be called if you are reconfiguring the player.
+     */
+    void completeConfiguration();
```
If it's just for reconfiguration, call it completeReconfiguration

kennytv commented on paper-api/src/main/java/io/papermc/paper/connection/PlayerGameConnection.java:
```diff
@@ -0,0 +1,21 @@
+     * This will, by default, cause the player to stay until their connection is released by
+     * {@link PlayerConfigurationConnection#completeConfiguration()}
+     */
+    void enterConfiguration();
```
reenterConfiguration?

kennytv commented on paper-api/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java:
```diff
@@ -301,6 +304,15 @@ public boolean isTransferred() {
+     * @return connection
+     */
+    @NotNull
+    public PlayerLoginConnection getPlayerLoginConnection() {
```
Just getConnection to be in line with the other connection getters in and outside of events

kennytv commented on paper-server/src/main/java/io/papermc/paper/connection/HorriblePlayerLoginEventHack.java:
```diff
@@ -0,0 +1,90 @@
+                
+                Please notify the following plugin developers: %s
+
+                Continued usage may cause instability in future builds.
```
This should be shortened/reworded a bit, maybe use a link to the javadocs instead of that "See ..." line:

```suggestion
                WARNING: Legacy PlayerLoginEvent usage detected!
                
                This event forces an alternative player loading path that is
                deprecated and will be removed in a future release.
                See the PlayerLoginEvent javadocs for more information.
                
                Please notify the following plugin developers: %s
```

kennytv commented on paper-api/src/main/java/org/bukkit/event/player/PlayerLoginEvent.java:
```diff
@@ -14,7 +16,12 @@
  * <i>entity</i> be postponed to the {@link PlayerJoinEvent} instead.
+ * @deprecated Use {@link PlayerConnectionValidateLoginEvent} to handle pre-login logic
+ * (e.g., authentication or ban checks), or {@link io.papermc.paper.event.player.PlayerServerFullCheckEvent} to allow
+ * players to bypass the server's maximum player limit.
```
This should also include info on why it's deprecated

kennytv reviewed (changes requested):
also ctrl+f and remove at least half of the mentions of the word "currently" (especially in event headers)💩

Machine-Maker commented on paper-api/src/main/java/io/papermc/paper/connection/PlayerConfigurationConnection.java:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @return the configuring player audience
+     */
+    Audience getConfiguringPlayer();
```
If its just an "Audience", how to I, in a callback where I only have the audience, get access to the methods I can use on the connection?

Owen1212055 committed:
Move Leaf's KeepAlive out of ServerPlayer

This needs to be moved out as there is no serverplayer in the configuration stage. Additionally, fix logging inconsistency.

Owen1212055 committed:
Fix resource packs sometimes throwing error, inline audience

Owen1212055 committed:
retain state

kennytv committed:
add back ServerGamePacketListenerImpl diff

kennytv committed:
Shorten warning

kennytv committed:
Fix owen javadoc
